### ISSUE TITLE
Introduce analyzer rule to warn against "nullable value type lifting"

### ISSUE THREAD

**Issue by drieseng** [meziantou/Meziantou.Analyzer#1533]
I've recently come across a bug in our code base that was related to "nullable value type lifting".
People tend to think that a null value for a nullable value type will "transformed" to the default value of that value type when used in arithmetic or relational operators, but that is surely not the case.

The following code fragment is a good illustration:
```c#
using System ;

int? nullableInt = null;
int x = 5;

Console.WriteLine((x + nullableInt) == null);  // Output: True
Console.WriteLine((x - nullableInt) == null);  // Output: True
Console.WriteLine(x < nullableInt);            // Output: False
Console.WriteLine(x > nullableInt);            // Output: False
Console.WriteLine(x == nullableInt);           // Output: False
```

It would be nice if an analyzer rule would warn against this.

**Comment by meziantou** [comment]
What would be the recommended syntax to replace `x + nullableInt`?

**Comment by drieseng** [comment]
In general I'd first perform a null / HasValue check:
```c#
if (nullableInt is { } value)
{
    Console.WriteLine(x + value);
}
```
I'm not sure a one-size-fits-all "fixer" can be added for this.

**Comment by meziantou** [comment]
While reviewing some code, I don't see any issues with using `nullable + value` that would lead to unexpected results. I’m also curious about why `x == value` could be considered problematic. Could you provide some real-world examples to clarify?

**Comment by drieseng** [comment]
Here's a (sort of) real-word example:
```c#
using System;

internal class Program
{
    private static void Main(string[] args)
    {
        var arrivalTime = new TimeSpan(10, 20, 0);
        var departureTime = new TimeSpan(10, 10, 00); 
        var bufferTime = GetBufferTime();

        if (arrivalTime + bufferTime > departureTime)
        {
            Console.WriteLine("Arrival time + Buffer time is greater than departure time.");
        }
        else
        {
            Console.WriteLine("Arrival time + Buffer time is not greater than departure time.");
        }
    }

    private static TimeSpan? GetBufferTime()
    {
        // Let's say that - depending on some business rule - this sometimes returns a TimeSpan and sometimes returns null.
        return null;
    }
}
```
Even though **arrivalTime** by itself is already greater than **departureTime**, adding a **null** value for **bufferTime** to it will result in **null** which is clearly not greater than **departureTime**.

Having a diagnostic rule for the "nullable value type lifting" would have alerted the developer to this. Depending on the requirements, he could have either have introduced a separate code path for when **bufferTime** is **null** or just used `bufferTime.GetValueOrDefault()`in the condition.

**Comment by meziantou** [comment]
I think this rule will have too many false-positives, so I'm not sure I want to implement it. However, I added [MA0240](https://github.com/meziantou/Meziantou.Analyzer/blob/main/docs/Rules/MA0240.md) yesterday which allow you to report syntax nodes matching custom expressions. I'll add data from the semantic model, so you should be able to write queries like `//AddExpression/*[@semantic:TypeConstructedFrom='System.Nullable`1']` to report all usages or `+` with a `Nullable<T>` argument.

**Comment by meziantou** [comment]
The latest version supports a subset of the semantic model, so you can use the following query in `BannedSyntaxes.txt`: ``//AddExpression/*[@semantic:TypeMetadataName='System.Nullable`1']; Do not add nullable values``.
